rnote-compose/src/shapes/arrow.rs

```diff
@@ -0,0 +1,212 @@
+
+use crate::helpers::{AabbHelpers, Vector2Helpers};
+use crate::shapes::ShapeBehaviour;
+use crate::transform::TransformBehaviour;
+
+use super::Line;
+
+type Radian = f64;
+
+/// All doc-comments of this file and [`ArrowBuilder`] rely on the following
+/// graphic:
+/// ```
+///         tip
+///         /|\
+///        / | \
+///       /  |  \
+///    lline |  rline
+///          |
+///          |
+///          |
+///         start
+/// ```
+/// Where `lline`, `tip`, `start` and `rline` represent a vector of the arrow.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(default, rename = "arrow")]
+pub struct Arrow {
+    /// The start of the arrow
+    pub start: na::Vector2<f64>,
+
+    /// The tip of the arow
+    pub tip: na::Vector2<f64>,
+}
+
+impl TransformBehaviour for Arrow {
+    fn translate(&mut self, offset: na::Vector2<f64>) {
+        self.start += offset;
+        self.tip += offset;
+    }
+
+    fn rotate(&mut self, angle: f64, center: na::Point2<f64>) {
+        let isometry = {
+            let mut isometry = na::Isometry2::identity();
+            isometry.append_rotation_wrt_point_mut(&na::UnitComplex::new(angle), &center);
+            isometry
+        };
+
+        self.start = (isometry * na::Point2::from(self.start)).coords;
+        self.tip = (isometry * na::Point2::from(self.tip)).coords;
+    }
+
+    fn scale(&mut self, scale: na::Vector2<f64>) {
+        self.start = self.start.component_mul(&scale);
+        self.tip = self.tip.component_mul(&scale);
+    }
+}
+
+impl ShapeBehaviour for Arrow {
+    fn bounds(&self) -> Aabb {
+        let (x_values, y_values) = {
+            let lline = self.compute_lline(None);
+            let rline = self.compute_rline(None);
+
+            let x_values = [lline[0], rline[0], self.start[0], self.tip[0]];
+            let y_values = [lline[1], rline[1], self.start[1], self.tip[1]];
+
+            (x_values, y_values)
+        };
+
+        let bottom_left_corner = {
+            let lowest_x = x_values.into_iter().reduce(f64::min).unwrap();
+            let lowest_y = y_values.into_iter().reduce(f64::min).unwrap();
+            na::Point2::new(lowest_x, lowest_y)
+        };
+
+        let top_right_corner = {
+            let highest_x = x_values.into_iter().reduce(f64::max).unwrap();
+            let highest_y = y_values.into_iter().reduce(f64::max).unwrap();
+            na::Point2::new(highest_x, highest_y)
+        };
+
+        AabbHelpers::new_positive(bottom_left_corner, top_right_corner)
+    }
+
+    fn hitboxes(&self) -> Vec<Aabb> {
+        let n_splits = super::hitbox_elems_for_shape_len((self.tip - self.start).norm());
+
+        self.split(n_splits)
+            .into_iter()
+            .map(|line| line.bounds())
+            .collect()
+    }
+}
+
+/// Contains helper-functions for the `ShapeBehaviour` implementation of
+/// `Arrow`.
+impl Arrow {
+    /// Splits the stem of the arrow into the given number of rectangles.
```

Comment:
this is a bit incorrect, no? I'd write essentially the same as for `Line`: "/// Splits itself given the no splits" or something similar

Reply:
Hm... I would say that this is correct:
![image](https://user-images.githubusercontent.com/50843046/230738978-d461b060-055e-4d43-a923-eab7f2fd75cb.png)

isn't that function computing the boxes "on" the stem?

Reply:
right, but then: "rectangles" -> "lines"

Reply:
ok